model/renderNativeWindow.cc

```diff
@@ -54,41 +63,127 @@ using namespace ecolab;
 #else
 
-  int throwOnXError(Display*, XErrorEvent* ev)
+  int throwOnXError(Display *, XErrorEvent *ev)
   {
     char errorMessage[256];
     XGetErrorText(ev->display, ev->error_code, errorMessage, sizeof(errorMessage));
     throw runtime_error(errorMessage);
   }
-  
-  inline cairo::SurfacePtr nativeWindowSurface(unsigned long window)
+
+  inline cairo::SurfacePtr createNativeWindowSurface(WindowInformation *wi)
   {
-    // ensure errors are thrown, rather than exit() being called
-    static bool errorHandlingSet=(XSetErrorHandler(throwOnXError), true);
-    XWindowAttributes wAttr;
-    auto display=XOpenDisplay(nullptr);
-    int err=XGetWindowAttributes(display, window, &wAttr);
-    if (err>1)
-      throw runtime_error("Invalid window: "+to_string(window));
-    cairo::SurfacePtr r(new cairo::Surface(cairo_xlib_surface_create(display,window,wAttr.visual,wAttr.width,wAttr.height),wAttr.width,wAttr.height));
-    cairo_surface_set_device_offset(r->surface(), -wAttr.x, -wAttr.y);
-    return r;
+    cairo::SurfacePtr childSurface(new cairo::Surface(cairo_xlib_surface_create(wi->display, wi->childWindowId, wi->wAttr.visual, wi->childWidth, wi->childHeight), wi->childWidth, wi->childHeight));
+    cairo_surface_set_device_offset(childSurface->surface(), -wi->wAttr.x, -wi->wAttr.y);
+    return childSurface;
   }
 #endif
-  
 
-  void RenderNativeWindow::renderToNativeWindow(unsigned long window)
+  WindowInformation::WindowInformation()
+  {
+  }
+  WindowInformation::~WindowInformation()
+  {
+    cout << "Delete called for window information" << endl;
+  }
+
+  void WindowInformation::initialize(unsigned long parentWin, int left, int top, int cWidth, int cHeight)
+  {
+    parentWindowId = parentWin;
+    offsetLeft = left;
+    offsetTop = top;
+
+    static bool errorHandlingSet = (XSetErrorHandler(throwOnXError), true);
+    display = XOpenDisplay(nullptr);
+    int err = XGetWindowAttributes(display, parentWin, &wAttr);
+    if (err > 1)
+      throw runtime_error("Invalid window: " + to_string(parentWin));
+
+    childWidth = wAttr.width - offsetLeft;
+    childHeight = wAttr.height - offsetTop;
+    
+    // Todo:: currently width, height parameters are not getting passed properly
+    // Eventually, we need those in order to ensure we don't need to worry about
+    // paddings, scrollbars etc
+    if(cWidth > 0) {
+      childWidth = min(childWidth, cWidth);
+    }
+    if(cHeight > 0) {
+      childHeight = min(childHeight, cHeight);
+    }
+
+    cout << childWidth << "::" << childHeight << "::" << offsetLeft << "::" << offsetTop << endl;
+    childWindowId = XCreateSimpleWindow(display, parentWin, offsetLeft, offsetTop, childWidth, childHeight, 0, 0, 0); //TODO:: Should we pass visual and attributes at the end?
+    XMapWindow(display, childWindowId);
+  }
+
+  void WindowInformation::copy(WindowInformation *winInfo)
+  {
+    this->parentWindowId = winInfo->parentWindowId;
+    this->childWindowId = winInfo->childWindowId;
```

Comment:
childWindowId is managing resources in just the same way as a pointer does, so this line leaks resources. The proper way to solve this is to apply RAII.